Add documention for the different error models including the models (#84)
PumasAI/PumasDocs.jl: docs/src/model_components/error_models.md

# Error models

> The error models in Pumas are formulated as the conditional distribution of the observed variable, ``Y``, conditionally on the random effects, ``η``. In the following, ``μ = \mathrm{E}(Y|η)`` is the individual prediction. The individual prediction ``μ`` generally dependents on the random effect ``η`` but the dependency will be suppressed in the notation for simplicity.

Reviewer: ```suggestion
The error models in Pumas are formulated as the conditional distribution of the observed variable, ``Y``, conditionally on the random effects, ``η``. In the following, ``μ = \mathrm{E}(Y|η)`` is the individual prediction. The individual prediction ``μ`` generally depends on the random effect ``η`` but the dependency will be suppressed in the notation for simplicity.
```

> A broad range of distributions are supported and a subset of the more common ones are presented below. A small code examples is presented for each of the error models presented as well as the conditional mean (individual prediction) and variance of each of the models.

Reviewer: ```suggestion
A broad range of distributions are supported and a subset of the more common ones is presented below. A small code example is presented for each of the error models presented as well as the conditional mean (individual prediction) and variance of each of the models.
```

## Gaussian models

> The Gaussian models are the traditional work horse models for non-linear mixed effects models. In Pumas, the Gaussian models are defined by using the `Normal` structure which is parameterized with the mean and the standard deviation.

Reviewer: ```suggestion
The Gaussian models are the traditional workhorse models for non-linear mixed effects models. In Pumas, the Gaussian models are defined by using the `Normal` structure which is parameterized with the mean and the standard deviation.
```

### Additive
```julia
@derived begin
  y ~ @. Normal(μ, σ)
end
```

```math
\mathrm{E}(Y|η) = μ \quad \mathrm{Var}(Y|η) = σ^2
```

The dispersion parameter of this model is ``σ^2`` which allows for estimation with the `Pumas.FOCE()` approximation.

### Proportional
```julia
@derived begin
  y ~ @. Normal(μ, μ*σ)
end
```

```math
\mathrm{E}(Y|η) = μ \quad \mathrm{Var}(Y|η) = (μσ)^2
```

The dispersion parameter of this model is ``(μσ)^2``, i.e. a function of ``μ``, and the model, therefore, does not allow for estimation with the `Pumas.FOCE()` approximation but requires the use of `Pumas.FOCEI()`.

> In some cases when using numerical ODE integration, it can be difficult to ensure that ``μ`` remains positive. In such cases, it usually help to write`Normal(μ, abs(μ)*σ)` or alternatively `Normal(μ, μ*σ + 1e-10)`.

Reviewer: ```suggestion
In some cases when using numerical ODE integration, it can be difficult to ensure that ``μ`` remains positive. In such cases, it usually helps to write`Normal(μ, abs(μ)*σ)` or alternatively `Normal(μ, μ*σ + 1e-10)`.
```

### Additive and proportional
```julia
@derived begin
  y ~ @. Normal(μ, sqrt(σ_add^2 + (μ*σ_prop)^2))
end
```

```math
\mathrm{E}(Y|η) = μ \quad \mathrm{Var}(Y|η) = σ_{add}^2 + (μσ_{prop})^2
```

The dispersion parameter of this model is ``σ_{add}^2 + (μσ_{prop})^2``, i.e. a function of ``μ``, and the model, therefore, does not allow for estimation with the `Pumas.FOCE()` approximation but requires the use of `Pumas.FOCEI()`.

### Log-normal (exponential)
```julia
@derived begin
  y ~ @. LogNormal(log(μ), σ)
end
```

```math
\mathrm{E}(Y|η) = \mathrm{E}\left(\exp\left(\log(μ) + \frac{σ^2}{2}\right)|η\right) \approx μ \quad \mathrm{Var}(Y|η) \approx (μσ)^2
```
for small values of ``σ`` so this model is very similar to the [Proportional](@ref) error model.

The dispersion parameter of this model is ``σ^2`` which allows for estimation with the `Pumas.FOCE()` approximation.

The log-normal model is equivalent to the following model

```julia
@derived begin
  logy ~ @. Normal(log(μ), σ)
end
```

> where `logy = log.(y)`. Notice that this is different from some other software packages where the log-normal model is equivalent to a proportional error model when using the first order approximation methods.

Reviewer: ```suggestion
where `logy = log.(y)`. Notice that this is different from some other software packages where the log-normal model is equivalent to a proportional error model when using the first-order approximation methods.
```

## Other continuous models
### Student's t

For data with some extreme observations, it is sometimes useful to use an error model with heavier tails such that Student's t distribution. The `LocationScale` structure is used to allow for location shift and scaling of one-parameter `TDist(ν)` distribution type.

```julia
@derived
  y ~ @. LocationScale(μ, σ, TDist(ν))
end
```

```math
\mathrm{E}(Y|η) = μ \quad \mathrm{Var}(Y|η) = σ^2 \quad \textrm{if } ν>2
```

### Gamma

> The Gamma model is traditionally used in the literature on generalized linear models, see [mccullagh1989generalized](@cite), for data where the the standard deviation is proportional to the mean which is same property that the Gaussian proportional error model has.

Reviewer: ```suggestion
The Gamma model is traditionally used in the literature on generalized linear models, see [mccullagh1989generalized](@cite), for data where the standard deviation is proportional to the mean which is the same property that the Gaussian proportional error model has.
```

> In contrast to the Gaussian proportional error model, the dispersion parameter of the Gamma model is ``ν^{-1}`` which not depend on the mean. This allows for estimation based on the `Pumas.FOCE()` approximation.

Reviewer: ```suggestion
In contrast to the Gaussian proportional error model, the dispersion parameter of the Gamma model is ``ν^{-1}`` which does not depend on the mean. This allows for estimation based on the `Pumas.FOCE()` approximation.
```

```julia
@derived begin
  y ~ @. Gamma(ν, μ/ν)
end
```

```math
\mathrm{E}(Y|η) = μ \quad \mathrm{Var}(Y|η) = \frac{μ}{ν}
```
which shows how the Gamma model is very similar to the Gaussian proportional error model. They share the first two moments when setting ``σ^2=ν^{-1}``.

## Models for discrete data
### Bernoulli (logistic regression)
```julia
@derived begin
  dv ~ @. Bernoulli(μ)
end
```

```math
\mathrm{E}(Y|η) = μ \quad \mathrm{Var}(Y|η) = μ(1 - μ)
```

The dispersion parameter of this model is ``1`` which allows for estimation with the `Pumas.FOCE()` approximation.

### Poisson
```julia
@derived begin
  y ~ @. Poisson(μ)
end
```

```math
\mathrm{E}(Y|η) = μ \quad \mathrm{Var}(Y|η) = μ
```

The dispersion parameter of this model is ``1`` which allows for estimation with the `Pumas.FOCE()` approximation.

### Negative binomial

> Notice that this distributions has many parametrizations in the literature.

Reviewer: ```suggestion
Notice that this distribution has many parametrizations in the literature.
```
